extension/goext/environment.go

```diff
@@ -162,6 +162,46 @@ func (nf *NullFloat) UnmarshalJSON(b []byte) error {
 	return nil
 }
 
+// IsDefined returns whether undefined value is defined
+func (ns *NullString) IsDefined() bool {
+	return ns != nil
+}
+
+// IsDefined returns whether undefined value is defined
+func (nb *NullBool) IsDefined() bool {
+	return nb != nil
+}
+
+// IsDefined returns whether undefined value is defined
+func (ni *NullInt) IsDefined() bool {
+	return ni != nil
+}
+
+// IsDefined returns whether undefined value is defined
+func (nf *NullFloat) IsDefined() bool {
+	return nf != nil
+}
+
+// Equals returns whether two undefined values are equal
+func (nf *NullFloat) Equals(rhv *NullFloat) bool {
+	return nf.IsDefined() && rhv.IsDefined() && nf.Valid && rhv.Valid && nf.Value == rhv.Value
```

Comment:
I am not sure about this logic.
Two undefined values should be equal in my opinion. The same applied to two invalid values.

Reply:
I've applied equality logic from JS. Please check comment above Equals functions to see logic table